dipy/reconst/fwdti.py

```diff
@@ -0,0 +1,876 @@
+    Diso : float, optional
+        Value of the free water isotropic diffusion. Default is set to 3e-3
+        $mm^{2}.s^{-1}$. Please ajust this value if you are assuming different
+        units of diffusion.
+    weighting: str, optional
+        the weighting scheme to use in considering the
+        squared-error. Default behavior is to use uniform weighting. Other
+        options: 'sigma' 'gmm'
+    sigma: float, optional
+        If the 'sigma' weighting scheme is used, a value of sigma needs to be
+        provided here. According to [Chang2005]_, a good value to use is
+        1.5267 * std(background_noise), where background_noise is estimated
+        from some part of the image known to contain no signal (only noise).
+    cholesky : bool, optional
+        If true it uses cholesky decomposition to insure that diffusion tensor
+        is positive define.
+        Default: False
+    f_transform : bool, optional
+        If true, the water volume fractions is converted during the convergence
+        procedure to ft = arcsin(2*f - 1) + pi/2, insuring f estimates between
+        0 and 1.
+        Default: True
+    jac : bool
+        Use the Jacobian? Default: False
+    mdreg : float, optimal
+        DTI's mean diffusivity regularization threshold. If standard DTI
+        diffusion tensor's mean diffusivity is almost near the free water
+        diffusion value, the diffusion signal is assumed to be only free water
+        diffusion (i.e. volume fraction will be set to 1 and tissue's diffusion
+        parameters are set to zero). Default md_reg is 2.7e-3 $mm^{2}.s^{-1}$
+        (corresponding to 90% of the free water diffusion value).
+
+    Returns
+    -------
+    fw_params : ndarray (x, y, z, 13)
+        Matrix containing in the dimention the free water model parameters in
+        the following order:
+            1) Three diffusion tensor's eigenvalues
+            2) Three lines of the eigenvector matrix each containing the
+               first, second and third coordinates of the eigenvector
+            3) The volume fraction of the free water compartment
+    S0 : ndarray (x, y, z)
+        The models estimate of the non diffusion-weighted signal S0.
+    """
+    # preparing data and initializing parameters
+    data = np.asarray(data)
+    data_flat = np.reshape(data, (-1, data.shape[-1]))
+
+    # Computing WLS DTI solution for MD regularization
+    dtiparams = dti_wls_fit(design_matrix, data_flat)
+    md = mean_diffusivity(dtiparams[..., :3])
+    cond = md > mdreg  # removal condition
+    data_cond = data_flat[~cond, :]
+
+    # Initializing fw_params according to selected initial guess
+    if np.any(fw_params) is None:
+        if np.any(S0) is None:
+            # If parameters and S0 are not given they both are estimated here
+            params_out, S0 = wls_fit_tensor(design_matrix, data_flat,
+                                            Diso=Diso)
```

Comment:
Test coverage is rather impressive already (at 98%), but this line is not being tested for some reason.